### ISSUE TITLE
PR #328 breaks queries when running inside AWS Lambda

### ISSUE THREAD

**Issue by joffotron** [snowflakedb/snowflake-connector-nodejs#364]
1. What version of NodeJS are you using (`node --version` and `npm --version`)?
14.21.1

3. What operating system and processor architecture are you using?
Amazon AWS Lambda, x86

4. What are the component versions in the environment (`npm list`)?
snowflake-sdk: 1.6.13

Was working in the version we were using previously.

5. What did you do?
Executing a largeish query using a bind of ~300kb JSON string 
```sql
INSERT INTO AMBERLANDING.DIRECT.TIMESCALE_USAGE_STAGING("USAGEID","USAGE_QUALITY","START","NETWORK_NAME","NMI","PRICE_QUALITY","NEM_TIME","SITE","CHANNEL_IDENTIFIER","NETWORK_REGION","TARIFF_INFO","USAGE_KWH","CUSTOMER_PRICE_CENTS","CHANNEL_TYPE","TARIFF_PRICE_CENTS","UPDATED_TIMESTAMP","AEMO_SPOT_PRICE_CENTS","ADJUSTED_SPOT_PRICE_CENTS") SELECT GET(value, 0),GET(value, 1),GET(value, 2),GET(value, 3),GET(value, 4),GET(value, 5),GET(value, 6),GET(value, 7),GET(value, 8),GET(value, 9),GET(value, 10),GET(value, 11),GET(value, 12),GET(value, 13),GET(value, 14),GET(value, 15),GET(value, 16),GET(value, 17) from TABLE(FLATTEN(PARSE_JSON(?)))
```

6. What did you expect to see?
The query successfully executing

8. What did you see instead?

The lambda runtime throws with this error `"errorType":"Error","errorMessage":"EROFS: read-only file system, open '1'","code":"EROFS","errno":-30`

Through our stacktraces, we have determined this query triggers a file write at https://github.com/snowflakedb/snowflake-connector-nodejs/blob/master/lib/connection/bind_uploader.js#L112

However, in the AWS Lambda runtime, the default current directory is `/var/task`, which is *readonly*. `/tmp` is available as a read-write filesystem however.

I humbly suggest changing the behaviour of this code to determine a safe place to write these temporary files to

**Comment by giacommobond007** [comment]
Thanks for posting this issue @joffotron, we are also encountering this with our lambdas.
We had to revert to version v1.6.13 to avoid this problem.

The performance benefits of BindUploader are admirable and a few improvements for stability across use-cases will be most welcome.

cc @SimbaGithub

**Comment by sfc-gh-dszmolka** [comment]
hi, thank you for raising this issue and providing your findings !
to work around this issue, i see two immediate options
* try and `chmod` this directory to a writable one if that's an option, at least this is what [AWS seems to recommend](https://docs.aws.amazon.com/lambda/latest/dg/troubleshooting-deployment.html)
* with the latest version of the driver (`v1.6.17`) now the `arrayBindingThreshold ` connection parameter is available; default value is `100000` , same as it was with prior versions, just it was not configurable. I think also specifying a higher value could probably help avoiding the problem because the driver will only try to use the improved performance path 'create files and upload them to a temporary stage before doing the actual `INSERT`) if the number of binds exceed a certain threshold , see above. Raising the threshold means the driver will still using the old path (like was available in 1.6.13 and before) . 
Let us know please if any of this helped.

Anyways I raised this to the team to consider improving the current behaviour.

**Comment by sfc-gh-dszmolka** [comment]
have another update here. [this change](https://github.com/snowflakedb/snowflake-connector-nodejs/pull/419/files#diff-c096e1194656cf2ec1329b578573677613fe6a4dd51cf123aa9af9e115c6342bL79-R81) (available with `v1.6.19` of the driver) might help in alleviating the situation in AWS Lambda / Google Cloud Functions / places with readonly default working directory.

if possible could someone who is affected please test and confirm if it's still an issue with the latest (1.6.19) version of the driver ?

**Comment by sfc-gh-dszmolka** [comment]
had a chance finally to test it myself. Function based on `snowflake-sdk@1.6.19` now uses `/tmp` in AWS Lambda, which is writable, and the script does not bail out with `EROFS` upon using the bind uploader. Example snippet from Lambda logs:
```
{"level":"DEBUG","message":"[3:29:23.2923 PM]: uploadFiles 0, @SYSTEM$BIND/ee248525-a830-413d-b57c-4656f07d2a14"}
{"level":"DEBUG","message":"[3:29:23.2923 PM]: Put=PUT file:///tmp/tmphMu6rB/1'@SYSTEM$BIND/ee248525-a830-413d-b57c-4656f07d2a14' overwrite=true auto_compress=false source_compression=gzip"}
..
{"level":"DEBUG","message":"[3:29:27.2927 PM]: uploadFiles done "}
{"level":"DEBUG","message":"[3:29:27.2927 PM]: stream on data"}
{"level":"DEBUG","message":"[3:29:27.2927 PM]: stream on end "}
{"level":"DEBUG","message":"[3:29:27.2927 PM]: all completed"}
{"level":"DEBUG","message":"[3:29:27.2927 PM]: Clean File=/tmp/tmphMu6rB/1"}
```
